File: core/src/main/java/lt/markmerkk/app/CameraHelper.java
package lt.markmerkk.app;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.math.Matrix4;
import org.jetbrains.annotations.NotNull;

public final class CameraHelper {
   private int viewportWidth;
   private int viewportHeight;
   private final int screenWidth;
   private final int screenHeight;
   private final float aspect;
   private final OrthographicCamera camera;
   private final int virtualWidth;
   private final int virtualHeight;

   public final int getViewportWidth() {
      return this.viewportWidth;
   }

   public final int getViewportHeight() {
      return this.viewportHeight;
   }

   @NotNull
   public final Matrix4 getCombine() {
      return camera.combined;
   }

   public CameraHelper(int virtualWidth, int virtualHeight) {
      this.virtualWidth = virtualWidth;
      this.virtualHeight = virtualHeight;
      this.camera = new OrthographicCamera();
      this.screenWidth = Gdx.graphics.getWidth();
      this.screenHeight = Gdx.graphics.getHeight();
      this.aspect = (float)(this.virtualWidth / this.virtualHeight);
      if((float)(this.screenWidth / this.screenHeight) >= this.aspect) {
         this.viewportHeight = this.virtualHeight;
         this.viewportWidth = this.viewportHeight * this.screenWidth / this.screenHeight;
      } else {
         this.viewportWidth = this.virtualWidth;
         this.viewportHeight = this.viewportWidth * this.screenHeight / this.screenWidth;
      }

      this.camera.setToOrtho(false, (float)this.viewportWidth, (float)this.viewportHeight);
   }
}

File: core/src/main/java/lt/markmerkk/app/CameraHelper.kt
package lt.markmerkk.app

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Matrix4
import org.slf4j.LoggerFactory

/**
 * @author mariusmerkevicius
 * @since 2016-06-04
 *
 * Responsible for moving user camera
 */
class CameraHelper(
        private val virtualWidth : Int,
        private val virtualHeight : Int
) {
    var viewportWidth: Int = 0
    var viewportHeight: Int = 0
    val combine: Matrix4
        get() = camera.combined

    private val screenWidth: Int
    private val screenHeight: Int
    private val aspect: Float

    private val camera: OrthographicCamera

    init {
        camera = OrthographicCamera()

        screenWidth = Gdx.graphics.width
        screenHeight = Gdx.graphics.height

        aspect = (virtualWidth / virtualHeight).toFloat()

        if (screenWidth / screenHeight >= aspect) {
            viewportHeight = virtualHeight
            viewportWidth = viewportHeight * screenWidth / screenHeight
        } else {
            viewportWidth = virtualWidth
            viewportHeight = viewportWidth * screenHeight / screenWidth
        }

        camera.setToOrtho(false, viewportWidth.toFloat(), viewportHeight.toFloat())
        logger.debug(String.format(
                "Screen: virtualWidth: %d x %d",
                virtualWidth,
                virtualHeight
        ))
        logger.debug(String.format(
                "Screen: screen: %d x %d",
                screenWidth,
                screenHeight
        ))
        logger.debug(String.format(
                "Screen: view port: %d x %d",
                viewportWidth,
                viewportHeight
        ))
    }

    fun update(x : Float, y : Float) {
        camera.position.set(x, y, 0f)
        camera.update()
    }

    companion object {
        val logger = LoggerFactory.getLogger(CameraHelper::class.java)
    }

}